superset-frontend/src/dashboard/util/crossFilters.ts

```diff
@@ -97,10 +111,13 @@ export const getCrossFiltersConfiguration = (
           },
         };
       }
+      const chartDataSource = charts[chartId].form_data.datasource;
       chartConfiguration[chartId].crossFilters.chartsInScope =
         isCrossFilterScopeGlobal(chartConfiguration[chartId].crossFilters.scope)
           ? globalChartConfiguration.chartsInScope.filter(
-              id => id !== Number(chartId),
+              id =>
+                id !== Number(chartId) &&
+                chartsByDataSource[chartDataSource].includes(id),
```

Comment:
Including optional chaining would likely help prevent undefined errors.

```suggestion
                chartsByDataSource[chartDataSource]?.has(id),
```